File: twitch.py
#!/usr/bin/env python3
import sys, os, errno, re, socket, random, time, atexit, imaplib, email

irc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
log_fh = None
log_name = ''
ts_re = re.compile(r'[tmi-]?sent-ts=(\d+)')

@atexit.register
def goodbye():
    if irc:
        irc.close()
    if log_fh:
        log_fh.close()

def update_log_fh(ts):
    global log_name, log_fh
    ts = time.strftime('logs/%Y/%B/%Y-%m-%d.txt', time.localtime(int(ts) / 1000))
    if ts != log_name:
        log_name = ts
        if log_fh:
            log_fh.close()
            log_fh = None

        if not os.path.exists(log_name):
            if not os.path.exists(os.path.dirname(log_name)):
                try:
                    os.makedirs(os.path.dirname(log_name), exist_ok=True)
                except OSError as exc:
                    if exc.errno != errno.EEXIST:
                        raise
            log_fh = open(log_name, 'wb')
        else:
            log_fh = open(log_name, 'ab')

def send(msg):
    irc.send("{}\r\n".format(msg).encode())

config = [x[:-1] for x in open('gmail.conf', 'r').readlines() if len(x) > 1]
with imaplib.IMAP4_SSL('imap.gmail.com') as mail:
    mail.login(config[0], config[1])
    mail.select("inbox")

    result, data = mail.uid('search', None, '(HEADER Subject "darksydephil just went live on Twitch") (UNSEEN)')
    if result == "OK" and data[0].decode("utf-8"):
        for uid in data[0].decode("utf-8").split(" "):
            result2, data2 = mail.uid('fetch', uid, '(RFC822)')
            if result2 == "OK" and data2:
                while True:
                    try:
                        user = 'justinfan' + ''.join(random.choice("0123456789") for _ in range(10))
                        irc.connect(("irc.chat.twitch.tv", 6667))
                        send("USER {} 0 0 :{}".format(user, user))
                        send("NICK {}".format(user))
                        send("CAP REQ :twitch.tv/membership")
                        send("CAP REQ :twitch.tv/tags")
                        send("CAP REQ :twitch.tv/commands")
                        send("JOIN #darksydephil")

                        split_msg_buf = None
                        while True:
                            try:
                                txt = irc.recv(512).decode('utf-8').split('\r\n')
                            except KeyboardInterrupt as e:
                                sys.exit(0)
                            except:
                                break

                            if split_msg_buf:
                                txt[0] = split_msg_buf + txt[0]
                                split_msg_buf = None
                            
                            if txt[-1]:
                                split_msg_buf = txt.pop()

                            for msg in txt:
                                if msg:
                                    if msg[:3] == "@ba": # All the interesting messages start with "@ba"
                                        ts = ts_re.findall(msg)
                                        if ts or not log_fh:
                                            update_log_fh(ts[1] if len(ts) >= 2 else ts[0])
                                        if log_fh:
                                            log_fh.write((msg + "\n").encode('utf-8'))
                                    elif msg[:4] == "PING":
                                        send("PO" + msg[2:])
                                    elif msg[:15] == "@msg-id=host_on":
                                        sys.exit(0)
                    except:
                        time.sleep(10)

File: twitch.rb
#!/usr/bin/env ruby
require "socket"
require "timeout"
require "FileUtils"
require "gmail"
require "thread"

# Extend IO to readlines without blocking
class IO
	def gets_nonblock
		@rlnb_buffer ||= ""
		ch = nil
		while ch = self.read_nonblock(1)
      @rlnb_buffer += ch
			if @rlnb_buffer[-1].ord == 10 && @rlnb_buffer[-2].ord == 13
				res  = @rlnb_buffer[0..-3]
				@rlnb_buffer = ""
				return res
			end
		end
	end
end

# Class to emit events
module Emitter
	def callbacks
		@callbacks ||= Hash.new { |h, k| h[k] = [] }
	end

	def on type, &block
		callbacks[type] << block
		self
	end

	def emit type, *args
		callbacks[type].each do |block|
			block.call(*args)
		end
	end
end

# Class to handle IRC stream and emit events
class Stream
	include Emitter
	attr_accessor :io, :buf

	def initialize serv, port
		@buf = []
		Timeout.timeout(5) do
		  @io = TCPSocket.new serv, port
		end
	rescue SocketError => e
		puts "ERROR! Failed to connect to #{serv}! #{e.message}"
		self.disconnect
	rescue Timeout::Error
		puts "ERROR! Connection to #{serv} timed out!"
		self.disconnect
	end

	def disconnect
		@io.puts 'QUIT'
		@io.close
	rescue Errno::EPIPE
	end

	def << data
		@buf << data
	end

	def write
		@buf.each do |x|
			@io.puts x
			emit :WROTE, x
		end
		@buf = []
	rescue EOFError, Errno::ECONNRESET
		emit :CLOSED
	end

	def read
		read = @io.gets_nonblock
		emit :READ, read
	rescue IO::WaitReadable
		emit :WAITING
	rescue EOFError, Errno::ECONNRESET
		emit :CLOSED
	end
end

# Class to handle IRC stream
class Bot
	attr_reader :stream

	def initialize stream
		@stream = stream
		@stream.on :CLOSED do
      stop
      @steam.disconnect
    end
	end

	def start
		@running = true
		tick while @running
	end

	def stop
		@running = false
	end

	def tick
		@stream.read
		@stream.write
		sleep 0.001
	end
end

bot  = nil
user = "justinfan" + (0...10).map { ('0'..'9').to_a[rand(10)] }.join
chan = "darksydephil"
out  = "./logs/%Y/%B/%Y-%m-%d.txt"
fh   = nil
fh_p = ""
tz   = "-07:00"
login, password = File.open('gmail.conf').readlines.map(&:rstrip)

begin
  Gmail.connect login, password do |gmail|
    gmail.inbox.find(:unread, :from => "no-reply@twitch.tv").each do |email|
      if email.message.subject == "darksydephil just went live on Twitch"
        email.read!

        sock = Stream.new "irc.chat.twitch.tv", 6667
        bot  = Bot.new sock

        sock << "USER #{user} 0 0 :#{user}"
        sock << "NICK #{user}"
        sock << "CAP REQ :twitch.tv/membership"
        sock << "CAP REQ :twitch.tv/tags"
        sock << "CAP REQ :twitch.tv/commands"
        sock << "JOIN ##{chan}"

				last_msg, timeout = Time.now, 900
				Thread.new do
					loop do
						bot.stop if Time.now - last_msg > timeout
						sleep 0.001
					end
				end

        sock.on :READ do |data|
          if data.start_with? "PING"
            data[1] = 'O'
            sock << data
          else
            time = Time.now.getlocal(tz)
						last_msg = Time.now
            path = time.strftime out
            unless path == fh_p
              fh_p = path
              fh.close unless fh.nil?
              fh_p_dn = File.dirname fh_p
              FileUtils.mkdir_p fh_p_dn unless File.directory? fh_p_dn
              fh = File.open fh_p, "a+"
            end
            fh.write data + "\n"
            bot.stop if data.start_with? "@msg-id=host_on"
          end
        end

        bot.start
      end
    end
  end
rescue SystemExit, Interrupt
	bot.stop unless bot.nil?
rescue EOFError, Exception
ensure
  fh.close unless fh.nil?
end
